clients/drcachesim/drpt2trace/pt2ir.h

```diff
@@ -0,0 +1,263 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* pt2ir: convert a PT raw trace to DynamoRIO's IR format. */
+
+#ifndef _PT2IR_H_
+#define _PT2IR_H_ 1
+
+/**
+ * @file drpt2trace/pt2ir.h
+ * @brief Offline PT raw trace converter.
+ */
+
+#include <string>
+#include <vector>
+#define DR_FAST_IR 1
+#include "dr_api.h"
+
+#ifndef IN
+#    define IN // nothing
+#endif
+#ifndef OUT
+#    define OUT // nothing
+#endif
+#ifndef INOUT
+#    define INOUT // nothing
+#endif
+
+/**
+ * The type of pt2ir_t::convert() return value.
+ */
+enum pt2ir_convert_status_t {
+    PT2IR_CONV_SUCCESS = 0,       /**< The conversion process is successful. */
+    PT2IR_CONV_SYNC_PACKET_ERROR, /**< The conversion process ends with a failure to sync
+                                   *   to the PSB packet.
+                                   */
+    PT2IR_CONV_HANDLE_SIDEBAND_EVENT_ERROR, /**< The conversion process ends with a
+                                             *   failure to handle a perf event.
+                                             */
+    PT2IR_CONV_GET_PENDING_EVENT_ERROR, /**< The conversion process ends with a failure to
+                                         *   get the pending event.
+                                         */
+    PT2IR_CONV_SET_IMAGE_ERROR, /**< The conversion process ends with a failure to set the
+                                 *   new image.
+                                 */
+    PT2IR_CONV_DECODE_NEXT_INSTR_ERROR /**< The conversion process ends with a failure to
+                                        *   decode the next intruction.
+                                        */
+};
+
+/**
+ * The struct pt2ir_config_t is a collection of one PT trace's configurations. drpt2trace
+ * and raw2trace can use it to construct pt2ir_t.
+ * XXX: Multiple PT raw traces for a process will share one kcore dump file. We may need a
+ * kernel image manager to avoid loading the same dump file many times.
+ */
+struct pt2ir_config_t {
+    /* The class pt2ir_t does not want to expose libipt to the upper layer. So we redefine
+     * some configurations of libipt decoder in pt_config. The libipt pt decoder requires
+     * these parameters. We can get these parameters by running
+     * libipt/scirpts/perf-get-opts.bash.
+     */
+    struct {
+        /* A cpu identifier.*/
```

Comment:
If these are all part of the library interface (which I assume they are, since there's a Doxygen comment for the outer struct here), these fields and sub-structs each need Doxygen comments: many are missing them as they have just a regular comment.

Reply:
Done.